### PR TITLE
Use grid-mean condensate for cloud optics

### PR DESCRIPTION
**Pull request by brhillman**
Use grid-mean condensate for cloud optics when not doing subcolumn sampling of cloud state. This adds a new yaml option to RRTMGP, `do_subcol_sampling`, to enable future support for implementing subcolumn sampling and to control whether or not we want to use grid-mean or in-cloud condensate amount to compute cloud optics. When not doing subcolumn sampling, we assume that each column is either entirely cloudy or entirely clear at a given level (i.e., layer cloud fraction is 0 or 1), and thus we want to use the grid-mean (total) cloud condensate amount to compute cloud optical properties. When doing subcolumn sampling of cloud state (not currently supported), we would want to use in-cloud values to compute cloud optical properties before assigning to subcolumns (i.e., using qc / cldfrac). The default for `do_subcol_sampling` is set to `false` for now, since subcolumn sampling is not yet implemented, and the code will throw an error if the user tries to select `do_subcol_sampling = true`. This change is non-BFB, since cloud properties are scaled differently for the computation of cloud optics.

[Non-B4B]

### CODE REVIEW

File: components/scream/src/physics/rrtmgp/atmosphere_radiation.cpp
Comment (AaronDonahue) [inline components/scream/src/physics/rrtmgp/atmosphere_radiation.cpp:556]: could we just initialize `cldfrac_tot` to 0 everywhere and then not need the `else` part of this conditional?

**Review by AaronDonahue — DISMISSED** [review]
Looks good, I had one question that was really just a style question, that won't change the functionality.

**Comment by AaronDonahue** [comment]
Note, this PR is failing `mappy` in SCREAMv1 tests with a DIFF.  Which is expected since it is non-BFB

File: components/scream/src/physics/rrtmgp/atmosphere_radiation.cpp
Comment (crterai) [inline components/scream/src/physics/rrtmgp/atmosphere_radiation.cpp:268]: What is the logic that determines the error message? If `m_do_subcol_sampling` is `not` true, then write out the message? That doesn't seem right, so I must be missing something.

**Review by crterai — COMMENTED** [review]
It looks like the PR will include a new namelist parameter of whether to use subcolumn sampling (MCICA), which by default is set to false. And when it's false, it will use all or nothing clouds to compute the radiation (based on whether cldfrc>0). Based on that, it looks good to me. 
I had a couple questions regarding where do_subcol_sampling is set to false by default and the logic to trigger the error message, but I'm less familiar with the C++ side of the model, so if Aaron thinks they look okay, I'm happy with this being merged.

File: components/scream/src/physics/rrtmgp/atmosphere_radiation.cpp
Comment (AaronDonahue) [reply components/scream/src/physics/rrtmgp/atmosphere_radiation.cpp:268]: ahhh, I think you are right @crterai , I'm surprised this would pass any tests without throwing an error right away.

File: components/scream/src/physics/rrtmgp/atmosphere_radiation.cpp
Comment (brhillman) [reply components/scream/src/physics/rrtmgp/atmosphere_radiation.cpp:556]: Yeah we probably could! I can do that but it does mean a separate kernel.

File: components/scream/src/physics/rrtmgp/atmosphere_radiation.cpp
Comment (brhillman) [reply components/scream/src/physics/rrtmgp/atmosphere_radiation.cpp:267]: Yup!

File: components/scream/src/physics/rrtmgp/atmosphere_radiation.cpp
Comment (AaronDonahue) [reply components/scream/src/physics/rrtmgp/atmosphere_radiation.cpp:268]: Ugg, now I'm so confused, but I think Ben is right.  So if `m_do_subcol_sampling` is FALSE than `not m_do_subcol_sampling` would be TRUE.  The REQUIRE does nothing if the input is TRUE.  So yeah, I take back my previous comment.

File: components/scream/src/physics/rrtmgp/atmosphere_radiation.cpp
Comment (brhillman) [reply components/scream/src/physics/rrtmgp/atmosphere_radiation.cpp:556]: It's not more involved, it's just a `memset(cldfrac_tot, 0.0)`, but my comment was that I think in either case this probably launches a kernel under the hood to do the "work" of setting to zero, especially since this is a device array/view.

File: components/scream/src/physics/rrtmgp/atmosphere_radiation.cpp
Comment (brhillman) [reply components/scream/src/physics/rrtmgp/atmosphere_radiation.cpp:268]:
```
#define IMPL_THROW(condition, msg, exception_type)                      \
  do {                                                                  \
    if ( ! (condition) ) {                                              \
      std::stringstream _ss_;                                           \
      _ss_ << __FILE__ << ":" << __LINE__ << ": FAIL:\n" << #condition; \
      _ss_ << "\n" << msg;                                              \
      throw exception_type(_ss_.str());                                 \
    }                                                                   \
  } while(0)

#define EKAT_REQUIRE_MSG(condition, msg)              IMPL_THROW(condition, msg, std::logic_error)
```

Comment (brhillman) [reply components/scream/src/physics/rrtmgp/atmosphere_radiation.cpp:268]: If this is too confusing, we could change the flag to something like `disable_subcol_sampling`, so the assertion would look like
```
EKAT_REQUIRE_MSG(disable_subcol_sampling, "ERROR! Need to disable subcol sampling for now.");
```
Or, we could take subcol sampling out of it and use a flag that just says `use_gridmean_clouds` and then the assertion would be
```
EKAT_REQUIRE_MSG(use_gridmean_clouds, "ERROR: need to use grid-mean clouds for cloud optics until MCICA sampling is implemented");
```

Comment (crterai) [reply components/scream/src/physics/rrtmgp/atmosphere_radiation.cpp:268]: Ahh, I understand now. It throws an error if the first condition is not satisfied. Then it looks good to me.

**Review by crterai — DISMISSED** [review]
My questions were resolved.

File: components/scream/src/physics/rrtmgp/atmosphere_radiation.cpp
Comment (AaronDonahue) [reply components/scream/src/physics/rrtmgp/atmosphere_radiation.cpp:268]: I think this is good, I think I just confused myself :-)

**Comment by AaronDonahue** [comment]
@brhillman . I think this branch just needs the whole `merge master in` treatment after the upstream merge went in.  Then I suspect we'll see passing everywhere except the expected DIFFs in SCREAMv1.

**Comment by brhillman** [comment]
Rebased.

**Review by crterai — APPROVED** [review]

**Review by AaronDonahue — APPROVED** [review]

**Comment by AaronDonahue** [comment]
The last time this ran through the AT it passed all tests except the expected DIFFs in the SCREAMv1 tests.  I think we can manually merge this, unless there are objections...

**Pull request merged**
